### ISSUE TITLE
haproxy 2.4.18 crashes

### ISSUE THREAD

**Issue by tarrychk** [haproxy/haproxy#1856]
### Detailed Description of the Problem

haproxy panics and kills the proxy process. Threads get stuck and wdt_handler then kills the process. We experience two issues which seem to have the same root cause.

When we had "option accept-invalid-http-request" enabled we had a lot of threads stuck in "proxy_capture_error" function, one of those then triggering wdt_handler and the ABRT:
```
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007f2014eaf859 in __GI_abort () at abort.c:79
#2  0x000055fd0804013c in ha_panic () at src/debug.c:333
#3  0x000055fd080a0587 in wdt_handler (sig=<optimized out>, si=<optimized out>, arg=<optimized out>) at src/wdt.c:122
#4  <signal handler called>
#5  0x000055fd08005a67 in proxy_capture_error (proxy=<optimized out>, is_back=<optimized out>, other_end=<optimized out>, target=<optimized out>, sess=<optimized out>, buf=<optimized out>, buf_ofs=<optimized out>, buf_out=<optimized out>, err_pos=<optimized out>,
    ctx=<optimized out>, show=<optimized out>) at src/proxy.c:2292
#6  0x000055fd07f5716a in h1_capture_bad_message (h1c=<optimized out>, h1s=h1s@entry=0x7f1fb8e11560, h1m=h1m@entry=0x7f1fb8e115a8, buf=buf@entry=0x7f1fdc224818) at src/mux_h1.c:1262
#7  0x000055fd07f60532 in h1_process_headers (max=<optimized out>, ofs=<optimized out>, buf=<optimized out>, htx=<optimized out>, h1m=<optimized out>, h1s=<optimized out>) at src/mux_h1.c:1448
#8  h1_process_input (h1c=h1c@entry=0x7f1fdc224800, buf=0x7f1fb8e11588, count=<optimized out>) at src/mux_h1.c:1576
#9  0x000055fd07f65602 in h1_process (h1c=h1c@entry=0x7f1fdc224800) at src/mux_h1.c:2771
#10 0x000055fd07f6652f in h1_io_cb (t=0x7f1fb8ddb230, ctx=0x7f1fdc224800, state=<optimized out>) at src/mux_h1.c:2930
#11 0x000055fd0805b74a in run_tasks_from_lists (budgets=budgets@entry=0x7f1fecb43600) at src/task.c:543
#12 0x000055fd0805c3a3 in process_runnable_tasks () at src/task.c:838
#13 0x000055fd080222ca in run_poll_loop () at src/haproxy.c:2626
#14 0x000055fd08022729 in run_thread_poll_loop (data=<optimized out>) at src/haproxy.c:2797
#15 0x00007f201556c609 in start_thread (arg=<optimized out>) at pthread_create.c:477
#16 0x00007f2014fac133 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

We commented out "option accept-invalid-http-request", but haproxy still is crashing, but now triggered by stktable_touch_local function.
```
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007f6c049f1859 in __GI_abort () at abort.c:79
#2  0x00005624fe93813c in ha_panic () at src/debug.c:333
#3  0x00005624fe998587 in wdt_handler (sig=<optimized out>, si=<optimized out>, arg=<optimized out>) at src/wdt.c:122
#4  <signal handler called>
#5  stktable_touch_local (t=t@entry=0x56250036a430, ts=ts@entry=0x7f6bc806b238, decrefcnt=decrefcnt@entry=0) at src/stick_table.c:455
#6  0x00005624fe91fe21 in stream_start_counters (ts=0x7f6bc806b238, t=<optimized out>) at include/haproxy/stream.h:210
#7  stream_track_stkctr (ts=0x7f6bc806b238, t=<optimized out>, ctr=<optimized out>) at include/haproxy/stream.h:225
#8  http_action_track_sc (rule=0x56250036d2d0, px=<optimized out>, sess=0x7f6bc14fda00, s=0x7f6bc386c930, flags=<optimized out>) at src/http_act.c:2129
#9  0x00005624fe843f6b in http_req_get_intercept_rule (px=px@entry=0x56250036a640, rules=rules@entry=0x56250036a680, s=s@entry=0x7f6bc386c930) at src/http_ana.c:2845
#10 0x00005624fe846f93 in http_process_req_common (s=0x7f6bc386c930, req=0x7f6bc386c940, an_bit=<optimized out>, px=0x56250036a640) at src/http_ana.c:369
#11 0x00005624fe86a0df in process_stream (t=<optimized out>, context=0x7f6bc386c930, state=<optimized out>) at src/stream.c:1912
#12 0x00005624fe9539ab in run_tasks_from_lists (budgets=budgets@entry=0x7f6be3de5600) at src/task.c:597
#13 0x00005624fe9543a3 in process_runnable_tasks () at src/task.c:838
#14 0x00005624fe91a2ca in run_poll_loop () at src/haproxy.c:2626
#15 0x00005624fe91a729 in run_thread_poll_loop (data=<optimized out>) at src/haproxy.c:2797
#16 0x00007f6c050ae609 in start_thread (arg=<optimized out>) at pthread_create.c:477
#17 0x00007f6c04aee133 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

### Expected Behavior

Threads shouldn't get stuck...

### Steps to Reproduce the Behavior

Issue is triggered by customer traffic, but no idea what exactly causes the crash. Without customer traffic haproxy runs stable. We were not able to reproduce the issue with test requests.

### Do you have any idea what may have caused this?

This issue seems to be triggered by specific traffic or load. It started unexpectedly at the same time and affected both haproxy instances we run in active/active setup at the same time. Without external traffic the issue does not happen, nor can we reproduce it with our test traffic.

### What is your configuration?

```haproxy
global
  log-tag haproxy-wgcs
  chroot /var/lib/haproxy
  maxconn 1199988
  server-state-file /etc/haproxy/lists/backends.state
  set-dumpable
  user haproxy
  group haproxy
  log /dev/log len 8192 local1
  log /dev/log len 8192 local2 err
  stats timeout 2m
  tune.maxrewrite 1024
  # we have seen requests and responses including large cookies/headers that failed
  tune.bufsize 18432
  tune.ssl.default-dh-param 2048
  ssl-default-bind-ciphers ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256
  ssl-default-bind-options ssl-min-ver TLSv1.2 no-tls-tickets
  ssl-default-server-ciphers ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256
  ssl-default-server-options no-sslv3 no-tlsv10 no-tlsv11 no-tls-tickets
  lua-load /etc/haproxy/bdcast.lua
  master-worker
  nbthread 40
  cpu-map 1/1-20 0-19
  cpu-map 1/21-40 44-63
  hard-stop-after 1h
  stats socket ipv4@127.0.0.1:9035 expose-fd listeners level admin
  stats socket /run/haproxy/admin.sock mode 600 expose-fd listeners level admin

listen stats
  bind 127.0.0.1:1936
  mode http
  stats enable
  stats uri /
  timeout connect 10s
  timeout client 1m
  timeout server 1m
  acl service_down nbsrv(lb-backend) lt 1
  monitor-uri /alive
  monitor fail if service_down
  http-request use-service prometheus-exporter if { path /metrics }

defaults
  log global
  load-server-state-from-file global
  option log-separate-errors
  maxconn 1199988
  timeout connect 5s
  timeout client 10m
  timeout client-fin 30s
  timeout server 10m
  timeout http-request 10s
  errorfile 400 /etc/haproxy/errors/400.http
  errorfile 403 /etc/haproxy/errors/403.http
  errorfile 408 /etc/haproxy/errors/408.http
  errorfile 500 /etc/haproxy/errors/500.http
  errorfile 502 /etc/haproxy/errors/502.http
  errorfile 503 /etc/haproxy/errors/503.http
  errorfile 504 /etc/haproxy/errors/504.http

backend st_err_global
  stick-table type ipv6 size 1000k expire 60s store http_err_rate(10s),http_req_rate(10s)

backend lb-backend
  fullconn 1199988
  timeout connect 25s
  timeout server 90m
  timeout http-request 25s
  mode http
  option accept-invalid-http-response
  acl blocked_host src -f /etc/haproxy/lists/ip_blocklist.lst
  http-request deny if blocked_host
  # block unresolvable known hosts
  acl unresolvable_host req.hdr(host) -f /etc/haproxy/lists/unresolvable_hosts.lst
  http-request deny if unresolvable_host
  # Rate limit the error case (4xx's)
  http-request track-sc0 src table st_err_global
  # Store the HTTP request rate and error rate in variables
  http-request set-var(req.req_rate) sc_http_req_rate(0)
  http-request set-var(req.err_rate) sc_http_err_rate(0)
  # compute percentage of error requests and compare against threshold
  # haproxy only supports integer math, no floats
  acl error_percentage_exceeded int(100),mul(req.err_rate),div(req.req_rate) gt 95
  acl error_rate_exceeded var(req.err_rate) gt 50
  http-request deny deny_status 429 if error_rate_exceeded error_percentage_exceeded
  # Call lua script to broadcast
  acl is_broadcast_url hdr(Host) -i broadcast.internal.com
  acl has_params_url urlp(rtparams) -m found
  http-request lua.broadcast_txn if is_broadcast_url has_params_url
  # stickyness configuration for load balancing and session validity (Host header, source IP)
  http-request set-header X-LB "%[req.hdr(Host),lower]_%[src]_%[src_port]"
  balance hdr(X-LB)
  # delete untrusted incoming XFF header
  http-request del-header X-Forwarded-For
  # do compression on outgoing traffic
  compression algo gzip
  compression type text/css text/html text/javascript application/javascript text/plain text/xml application/json
  hash-type consistent
  # configure thresholds for filter nodes
  default-server inter 3s fastinter 1s downinter 500ms fall 3 rise 2 slowstart 10s
  server node00 SRV1:8080 source SRC1 check maxconn 33333 send-proxy
  server node01 SRV1:8080 source SRC2 check maxconn 33333 send-proxy
  server node02 SRV1:8080 source SRC2 check maxconn 33333 send-proxy

frontend static
  bind SRC1:8080
  timeout client 3h
  timeout http-request 25s
  http-response set-header Strict-Transport-Security "max-age=31536000" if { ssl_fc }
  mode http
  #option accept-invalid-http-request
  option dontlognull
  acl backend_down nbsrv(lb-backend-mwg) lt 1
  monitor-uri /alive
  monitor fail if backend_down
  stick-table type string size 1k expire 1m store http_req_rate(1s)
  acl is_broadcast_url hdr(Host) -i broadcast.internal.com
  http-request track-sc0 hdr(Host) if is_broadcast_url
  http-request deny if { sc0_http_req_rate  gt 1 } is_broadcast_url
  default_backend lb-backend
```

### Output of `haproxy -vv`

```plain
HAProxy version 2.4.18-1ppa1~focal 2022/07/27 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2026.
Known bugs: http://www.haproxy.org/bugs/bugs-2.4.18.html
Running on: Linux 4.15.0-163-generic #171-Ubuntu SMP Fri Nov 5 11:55:11 UTC 2021 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -O2 -fdebug-prefix-map=/build/haproxy-96Se88/haproxy-2.4.18=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_SYSTEMD=1 USE_PROMEX=1
  DEBUG   =

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -CLOSEFROM -ZLIB +SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL -PROCCTL +THREAD_DUMP -EVPORTS -OT -QUIC +PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=64).
Built with OpenSSL version : OpenSSL 1.1.1f  31 Mar 2020
Running on OpenSSL version : OpenSSL 1.1.1f  31 Mar 2020
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.34 2019-11-21
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 9.4.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2       flags=HTX|CLEAN_ABRT|HOL_RISK|NO_UPG
            fcgi : mode=HTTP       side=BE        mux=FCGI     flags=HTX|HOL_RISK|NO_UPG
       <default> : mode=HTTP       side=FE|BE     mux=H1       flags=HTX
              h1 : mode=HTTP       side=FE|BE     mux=H1       flags=HTX|NO_UPG
       <default> : mode=TCP        side=FE|BE     mux=PASS     flags=
            none : mode=TCP        side=FE|BE     mux=PASS     flags=NO_UPG

Available services : prometheus-exporter
Available filters :
        [SPOE] spoe
        [CACHE] cache
        [FCGI] fcgi-app
        [COMP] compression
        [TRACE] trace
```

### Last Outputs and Backtraces

```plain
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
        set = {__val = {18446744067195533847, 0 <repeats 13 times>, 1179670597, 0}}
        pid = <optimized out>
        tid = <optimized out>
        ret = <optimized out>
#1  0x00007f6c049f1859 in __GI_abort () at abort.c:79
        save_stage = 1
        act = {__sigaction_handler = {sa_handler = 0xb, sa_sigaction = 0xb}, sa_mask = {__val = {4294967300, 18446744067267100671, 18446744073709551615 <repeats 12 times>, 140101917840196, 18446744073709551615}}, sa_flags = 18432, sa_restorer = 0x7f6bc001e5b0}
        sigs = {__val = {32, 0 <repeats 15 times>}}
#2  0x00005624fe93813c in ha_panic () at src/debug.c:333
No locals.
#3  0x00005624fe998587 in wdt_handler (sig=<optimized out>, si=<optimized out>, arg=<optimized out>) at src/wdt.c:122
        n = <optimized out>
        p = <optimized out>
        thr = <optimized out>
#4  <signal handler called>
No locals.
#5  stktable_touch_local (t=t@entry=0x56250036a430, ts=ts@entry=0x7f6bc806b238, decrefcnt=decrefcnt@entry=0) at src/stick_table.c:455
        __pl_l = 0x56250036a630
        __pl_r = <optimized out>
        expire = 406635617
#6  0x00005624fe91fe21 in stream_start_counters (ts=0x7f6bc806b238, t=<optimized out>) at include/haproxy/stream.h:210
        ptr = <optimized out>
        ptr = <optimized out>
        __pl_r = <optimized out>
        __pl_l = <optimized out>
        ret = <optimized out>
        __pl_l = <optimized out>
        __pl_r = <optimized out>
        __pl_i = <optimized out>
        ret = <optimized out>
        __pl_i = <optimized out>
#7  stream_track_stkctr (ts=0x7f6bc806b238, t=<optimized out>, ctr=<optimized out>) at include/haproxy/stream.h:225
No locals.
#8  http_action_track_sc (rule=0x56250036d2d0, px=<optimized out>, sess=0x7f6bc14fda00, s=0x7f6bc386c930, flags=<optimized out>) at src/http_act.c:2129
        t = <optimized out>
        ts = 0x7f6bc806b238
        key = <optimized out>
        ptr1 = 0x0
        ptr2 = 0x0
        ptr3 = 0x0
        ptr4 = 0x0
        ptr5 = 0x0
        ptr6 = 0x0
        opt = <optimized out>
        __pl_r = <optimized out>
        __pl_i = <optimized out>
        ret = <optimized out>
        __pl_i = <optimized out>
#9  0x00005624fe843f6b in http_req_get_intercept_rule (px=px@entry=0x56250036a640, rules=rules@entry=0x56250036a680, s=s@entry=0x7f6bc386c930) at src/http_ana.c:2845
        sess = <optimized out>
        txn = 0x7f6bc14fd4c0
        rule = 0x56250036d2d0
        rule_ret = HTTP_RULE_RES_CONT
        act_opts = 2
#10 0x00005624fe846f93 in http_process_req_common (s=0x7f6bc386c930, req=0x7f6bc386c940, an_bit=<optimized out>, px=0x56250036a640) at src/http_ana.c:369
        sess = <optimized out>
        txn = <optimized out>
        msg = <optimized out>
        htx = <optimized out>
        rule = <optimized out>
        verdict = <optimized out>
        conn = 0x7f6bc14fd930
#11 0x00005624fe86a0df in process_stream (t=<optimized out>, context=0x7f6bc386c930, state=<optimized out>) at src/stream.c:1912
        srv = <optimized out>
        s = 0x7f6bc386c930
        sess = <optimized out>
        rqf_last = <optimized out>
        rpf_last = 2147483648
        rq_prod_last = 8
        rq_cons_last = 0
        rp_cons_last = 8
        rp_prod_last = 0
        req_ana_back = <optimized out>
        req = 0x7f6bc386c940
        res = 0x7f6bc386c9a0
        si_f = 0x7f6bc386cbe0
        si_b = 0x7f6bc386cc38
        rate = <optimized out>
#12 0x00005624fe9539ab in run_tasks_from_lists (budgets=budgets@entry=0x7f6be3de5600) at src/task.c:597
        process = <optimized out>
        tl_queues = 0x5624feb095c0 <task_per_thread+2624>
        t = 0x7f6bc386cda0
        budget_mask = <optimized out>
        profile_entry = 0x0
        done = <optimized out>
        queue = <optimized out>
        state = <optimized out>
        ctx = <optimized out>
#13 0x00005624fe9543a3 in process_runnable_tasks () at src/task.c:838
        tt = 0x5624feb09580 <task_per_thread+2560>
        lrq = <optimized out>
        grq = <optimized out>
        t = <optimized out>
        max = {0, 4, 2, 1}
        max_total = <optimized out>
        tmp_list = <optimized out>
        queue = <optimized out>
        max_processed = <optimized out>
        lpicked = <optimized out>
        gpicked = <optimized out>
        heavy_queued = 1
        budget = <optimized out>
#14 0x00005624fe91a2ca in run_poll_loop () at src/haproxy.c:2626
        next = <optimized out>
        wake = <optimized out>
#15 0x00005624fe91a729 in run_thread_poll_loop (data=<optimized out>) at src/haproxy.c:2797
        ptaf = <optimized out>
        ptif = <optimized out>
        ptdf = <optimized out>
        ptff = <optimized out>
        init_left = 0
        init_mutex = {__data = {__lock = 0, __count = 0, __owner = 0, __nusers = 0, __kind = 0, __spins = 0, __elision = 0, __list = {__prev = 0x0, __next = 0x0}}, __size = '\000' <repeats 39 times>, __align = 0}
        init_cond = {__data = {{__wseq = 243, __wseq32 = {__low = 243, __high = 0}}, {__g1_start = 231, __g1_start32 = {__low = 231, __high = 0}}, __g_refs = {0, 0}, __g_size = {0, 0}, __g1_orig_size = 24, __wrefs = 0, __g_signals = {0, 0}},
          __size = "\363\000\000\000\000\000\000\000\347", '\000' <repeats 23 times>, "\030", '\000' <repeats 14 times>, __align = 243}
#16 0x00007f6c050ae609 in start_thread (arg=<optimized out>) at pthread_create.c:477
        ret = <optimized out>
        pd = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140101361272576, 1682253566209805731, 140727578389694, 140727578389695, 140727578389696, 140101361227712, -1616978048491336285, -1621125848105759325}, mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x0, 0x0}, data = {
              prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = 0
#17 0x00007f6c04aee133 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

### Additional Information

We are using haproxy packages from Vincent Bernat PPA (https://launchpad.net/~vbernat/+archive/ubuntu/haproxy-2.4).

**Comment by wtarreau** [comment]
Thanks. One shows an infinite wait on the proxy lock, the other one on the table lock. I'm wondering if there could be an AB-BA situation here. What's the last version that didn't exhibit this problem ?

**Comment by capflam** [comment]
I don't think it is an AB-BA issue because the crashes seem to be always exclusively on the proxy lock or on the table lock depending on the `accept-invalid-http-request` option. However it seems to be related to the load. Could you provide `lscpu  --extended` output ? In addition, a full backtrace on all threads (`t a a bt ful` on gdbl) could help.

**Comment by wtarreau** [comment]
t a a bt full will definitely tell us if other code paths are conflicting. If it's related to the load, it shouldn't loop forever without progressing on a lock like this (unless it's running on an EPYC where this tends to happen from time to time).

**Comment by capflam** [comment]
I agree, That's why I asked for the cpu info.

**Comment by tarrychk** [comment]
> What's the last version that didn't exhibit this problem ?

We saw it on 2.4.18 exclusively, before that we were running 2.4.15. We didn't experience any issues since updating to 2.6.5.
That said, 2.4.18 was running fine for 2 weeks before the crashes happened, and we have multiple 2.4.18 in other locations which don't show any issues, either.

> lscpu --extended

CPU NODE SOCKET CORE L1d:L1i:L2:L3 ONLINE MAXMHZ    MINMHZ
0   0    0      0    0:0:0:0       yes    3600.0000 1200.0000
1   0    0      1    1:1:1:0       yes    3600.0000 1200.0000
2   0    0      2    2:2:2:0       yes    3600.0000 1200.0000
3   0    0      3    3:3:3:0       yes    3600.0000 1200.0000
4   0    0      4    4:4:4:0       yes    3600.0000 1200.0000
5   0    0      5    5:5:5:0       yes    3600.0000 1200.0000
6   0    0      6    6:6:6:0       yes    3600.0000 1200.0000
7   0    0      7    7:7:7:0       yes    3600.0000 1200.0000
8   0    0      8    8:8:8:0       yes    3600.0000 1200.0000
9   0    0      9    9:9:9:0       yes    3600.0000 1200.0000
10  0    0      10   10:10:10:0    yes    3600.0000 1200.0000
11  0    0      11   11:11:11:0    yes    3600.0000 1200.0000
12  0    0      12   12:12:12:0    yes    3600.0000 1200.0000
13  0    0      13   13:13:13:0    yes    3600.0000 1200.0000
14  0    0      14   14:14:14:0    yes    3600.0000 1200.0000
15  0    0      15   15:15:15:0    yes    3600.0000 1200.0000
16  0    0      16   16:16:16:0    yes    3600.0000 1200.0000
17  0    0      17   17:17:17:0    yes    3600.0000 1200.0000
18  0    0      18   18:18:18:0    yes    3600.0000 1200.0000
19  0    0      19   19:19:19:0    yes    3600.0000 1200.0000
20  0    0      20   20:20:20:0    yes    3600.0000 1200.0000
21  0    0      21   21:21:21:0    yes    3600.0000 1200.0000
22  1    1      22   22:22:22:1    yes    3600.0000 1200.0000
23  1    1      23   23:23:23:1    yes    3600.0000 1200.0000
24  1    1      24   24:24:24:1    yes    3600.0000 1200.0000
25  1    1      25   25:25:25:1    yes    3600.0000 1200.0000
26  1    1      26   26:26:26:1    yes    3600.0000 1200.0000
27  1    1      27   27:27:27:1    yes    3600.0000 1200.0000
28  1    1      28   28:28:28:1    yes    3600.0000 1200.0000
29  1    1      29   29:29:29:1    yes    3600.0000 1200.0000
30  1    1      30   30:30:30:1    yes    3600.0000 1200.0000
31  1    1      31   31:31:31:1    yes    3600.0000 1200.0000
32  1    1      32   32:32:32:1    yes    3600.0000 1200.0000
33  1    1      33   33:33:33:1    yes    3600.0000 1200.0000
34  1    1      34   34:34:34:1    yes    3600.0000 1200.0000
35  1    1      35   35:35:35:1    yes    3600.0000 1200.0000
36  1    1      36   36:36:36:1    yes    3600.0000 1200.0000
37  1    1      37   37:37:37:1    yes    3600.0000 1200.0000
38  1    1      38   38:38:38:1    yes    3600.0000 1200.0000
39  1    1      39   39:39:39:1    yes    3600.0000 1200.0000
40  1    1      40   40:40:40:1    yes    3600.0000 1200.0000
41  1    1      41   41:41:41:1    yes    3600.0000 1200.0000
42  1    1      42   42:42:42:1    yes    3600.0000 1200.0000
43  1    1      43   43:43:43:1    yes    3600.0000 1200.0000
44  0    0      0    0:0:0:0       yes    3600.0000 1200.0000
45  0    0      1    1:1:1:0       yes    3600.0000 1200.0000
46  0    0      2    2:2:2:0       yes    3600.0000 1200.0000
47  0    0      3    3:3:3:0       yes    3600.0000 1200.0000
48  0    0      4    4:4:4:0       yes    3600.0000 1200.0000
49  0    0      5    5:5:5:0       yes    3600.0000 1200.0000
50  0    0      6    6:6:6:0       yes    3600.0000 1200.0000
51  0    0      7    7:7:7:0       yes    3600.0000 1200.0000
52  0    0      8    8:8:8:0       yes    3600.0000 1200.0000
53  0    0      9    9:9:9:0       yes    3600.0000 1200.0000
54  0    0      10   10:10:10:0    yes    3600.0000 1200.0000
55  0    0      11   11:11:11:0    yes    3600.0000 1200.0000
56  0    0      12   12:12:12:0    yes    3600.0000 1200.0000
57  0    0      13   13:13:13:0    yes    3600.0000 1200.0000
58  0    0      14   14:14:14:0    yes    3600.0000 1200.0000
59  0    0      15   15:15:15:0    yes    3600.0000 1200.0000
60  0    0      16   16:16:16:0    yes    3600.0000 1200.0000
61  0    0      17   17:17:17:0    yes    3600.0000 1200.0000
62  0    0      18   18:18:18:0    yes    3600.0000 1200.0000
63  0    0      19   19:19:19:0    yes    3600.0000 1200.0000
64  0    0      20   20:20:20:0    yes    3600.0000 1200.0000
65  0    0      21   21:21:21:0    yes    3600.0000 1200.0000
66  1    1      22   22:22:22:1    yes    3600.0000 1200.0000
67  1    1      23   23:23:23:1    yes    3600.0000 1200.0000
68  1    1      24   24:24:24:1    yes    3600.0000 1200.0000
69  1    1      25   25:25:25:1    yes    3600.0000 1200.0000
70  1    1      26   26:26:26:1    yes    3600.0000 1200.0000
71  1    1      27   27:27:27:1    yes    3600.0000 1200.0000
72  1    1      28   28:28:28:1    yes    3600.0000 1200.0000
73  1    1      29   29:29:29:1    yes    3600.0000 1200.0000
74  1    1      30   30:30:30:1    yes    3600.0000 1200.0000
75  1    1      31   31:31:31:1    yes    3600.0000 1200.0000
76  1    1      32   32:32:32:1    yes    3600.0000 1200.0000
77  1    1      33   33:33:33:1    yes    3600.0000 1200.0000
78  1    1      34   34:34:34:1    yes    3600.0000 1200.0000
79  1    1      35   35:35:35:1    yes    3600.0000 1200.0000
80  1    1      36   36:36:36:1    yes    3600.0000 1200.0000
81  1    1      37   37:37:37:1    yes    3600.0000 1200.0000
82  1    1      38   38:38:38:1    yes    3600.0000 1200.0000
83  1    1      39   39:39:39:1    yes    3600.0000 1200.0000
84  1    1      40   40:40:40:1    yes    3600.0000 1200.0000
85  1    1      41   41:41:41:1    yes    3600.0000 1200.0000
86  1    1      42   42:42:42:1    yes    3600.0000 1200.0000
87  1    1      43   43:43:43:1    yes    3600.0000 1200.0000

**Comment by tarrychk** [comment]
[full-bt.txt](https://github.com/haproxy/haproxy/files/9534959/full-bt.txt)
I uploaded the requested full backtrace. I removed headers from the backtrace, let me know if you need that information for analysis.

**Comment by tarrychk** [comment]
[full-bt-2.txt](https://github.com/haproxy/haproxy/files/9535052/full-bt-2.txt)

**Comment by wtarreau** [comment]
Very useful, thank you. A lot of error captures at the same time, but in addition, threads 39, 35 and 26 are calling free() apparently under the proxy's lock! The call sleeps in the libc so that doesn't help at all!
One day we should really try to get rid of that pointless lock, it was used to force to serialize with another part that relies on it (I don't remember which one though). Could you please try to apply this patch to your version ?
```
diff --git a/src/proxy.c b/src/proxy.c
index 1ee483754..4df8de2cd 100644
--- a/src/proxy.c
+++ b/src/proxy.c
@@ -2295,8 +2295,8 @@ void proxy_capture_error(struct proxy *proxy, int is_back,
        } else {
                es = HA_ATOMIC_XCHG(&proxy->invalid_req, es);
        }
-       free(es);
        HA_RWLOCK_WRUNLOCK(PROXY_LOCK, &proxy->lock);
+       ha_free(&es);
 }
 
 /* Configure all proxies which lack a maxconn setting to use the global one by
```

Do not hesitate to let us know if you don't know how to proceed or don't feel at ease with modifying and rebuilding the code!

**Comment by tarrychk** [comment]
We can certainly apply the patch, but currently all instances run without issue (and most of them have done so for quite some time), so we can't really reproduce the crashes at will.

**Comment by tarrychk** [comment]
Looks like I messed up the second full backtrace and merged both traces into the same file...
I'll attach the clean bt for the stktable lockups again:
[full-bt-stktable.txt](https://github.com/haproxy/haproxy/files/9545178/full-bt-stktable.txt)

**Comment by wtarreau** [comment]
Yeah I noticed, but don't worry, it just confirms the first one, i.e. many threads waiting for the tabke lock that is held by another one waiting on a lock in free() while other threads are heavily hammering this lock in malloc().

I think I'll merge the patch anyway, as there's no point keeping it out.

Thanks!

**Comment by wtarreau** [comment]
Now merged into 2.7-dev.
